File: 백준/Silver/11660. 구간 합 구하기 5/구간 합 구하기 5.js
const readline = require('readline').createInterface({
  input: process.stdin,
  output: process.stdout
});

let input = [];
let N, M;

readline.on('line', (line) => {
  input.push(line);
}).on('close', () => {
  [N, M] = input[0].split(' ').map(Number);
  const result = solution(input.slice(1));
  console.log(result.join('\n'));
  process.exit();
});

const solution = (input) => {
  let table = input.slice(0, N);
  let result = [];
  let testcase = input.slice(N);
  let DP = Array.from({ length: N }, () => new Array(N).fill(0));

  for (let i = 0; i < N; i++) {
    table[i] = table[i].split(' ').map(Number);
  }

  for (let i = 0; i < N; i++) {
    for (let j = 0; j < N; j++) {
      DP[i][j] = table[i][j] + (j-1 >= 0 ? DP[i][j-1] : 0);
    }
  }

  for (let i = 1; i < N; i++) {
    for (let j = 0; j < N; j++) {
      DP[i][j] = DP[i-1][j] + DP[i][j];
    }
  }

  for (const test of testcase) {
    const [x1, y1, x2, y2] = test.split(' ').map(Number);
    const minus = (y1-2 >= 0 ? DP[x2-1][y1-2] : 0)+(x1-2 >= 0 ? DP[x1-2][y2-1] : 0);
    const plus = x1-2 >= 0 && y1-2 >= 0 ? DP[x1-2][y1-2] : 0;
    let sum = DP[x2-1][y2-1] - minus + plus;
    result.push(sum);
  }

  return result;
}

File: 백준/Silver/11660. 구간 합 구하기 5/구간 합 구하기 5.py
import sys
input = sys.stdin.readline

n, m = map(int, input().split())
matrix = []
case = []

for _ in range(n):
  matrix.append(list(map(int, input().split())))
for _ in range(m):
  case.append(list(map(int, input().split())))

dp = [[0 for _ in range(n)] for _ in range(n)]

for i in range(n):
  for j in range(n):
    if j-1 >= 0:
      dp[i][j] = dp[i][j-1]+matrix[i][j]
    else:
      dp[i][j] = matrix[i][j]

for i in range(n):
  for j in range(n):
    if i-1 >= 0:
      dp[i][j] = dp[i-1][j]+dp[i][j]

for x1, y1, x2, y2 in case:
  x1 -= 1; y1 -= 1; x2 -= 1; y2 -= 1
  all = dp[x2][y2]
  minus1 = dp[x1-1][y2] if x1-1 >= 0 else 0
  minus2 = dp[x2][y1-1] if y1-1 >= 0 else 0
  plus = dp[x1-1][y1-1] if x1-1 >= 0 and y1-1 >= 0 else 0
  sum = all-minus1-minus2+plus
  print(sum)
